### PR TITLE
feat: record MCP actions in audit logs

### PR DESCRIPTION
**Pull request by gonbaum**
### What does it do?

Records actions performed through the MCP server in the EE audit logs.

- The MCP handler (`handlePost.ts`) sets the authenticated admin user (the token owner) on the request state and marks the action's origin via `ctx.state.auditSource = 'mcp'`. The route type is left untouched — MCP is admin-authenticated but is not an admin route, so we don't fabricate one.
- The audit-logs guard (`lifecycles.ts`) records an event when it's a normal admin action **or** carries the MCP marker, and tags every recorded action's `payload.origin` with its origin (`'mcp'` or `'admin'`).

Backwards compatible (new JSON key in the payload, nothing else changes) and reversible in a few lines. Reads are not audited (they don't emit `entry.*` events).

### Why is it needed?

MCP-triggered writes were silently dropped by the audit-logs guard: MCP requests don't come through an admin route (`route.info.type !== 'admin'`) and don't set `ctx.state.user`, so the guard discarded their events even though the Document Service emitted them.

### How to test it?

Requires an Enterprise license that **includes the `audit-logs` feature** (Growth does not).

1. Start `examples/getstarted` with `mcp.enabled` and `features.future.adminTokens` on (already set locally), `PORT=1338 yarn develop`.
2. Create a Full-access **Admin API Token** in Settings → Admin API Tokens.
3. Via MCP (`POST /mcp`, initialize then `tools/call`): create an entry, list entries, delete the entry.
4. In Settings → Audit Logs: the create and delete appear with `payload.origin === "mcp"`; the list/read produces **no** log. Admin-panel actions keep `payload.origin === "admin"`.

Automated coverage: `STRAPI_LICENSE=<license-with-audit-logs> yarn test:api -- tests/api/core/mcp/mcp-audit-logs.test.api.ts` (admin create audited, MCP create audited, MCP read not audited, MCP delete audited).

### Related issue(s)/PR(s)

[EE-39](https://linear.app/strapi/issue/EE-39/audit-logs-compatibility)

### CODE REVIEW

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    action,
+    subject: MODEL_UID,
+    conditions: [],
+    properties: {},
+  });
+
+  // entry.* events fire on a microtask after commit; let the subscriber run.
+  const flushEvents = async () => {
+    await new Promise((resolve) => setTimeout(resolve, 200));
+  };
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts]: **P2 · Confidence: confirmed**

The fixed 200 ms sleep makes this regression test timing-dependent. The audit insert runs from an async post-commit callback whose promise is not awaited, so a slow CI runner or database can legitimately exceed this delay.

Please replace the delay with bounded polling for the expected audit row: query immediately, retry at a short interval, return as soon as the row appears, and fail after a finite timeout. For example:

```ts
const waitForAuditLog = async (action: string, timeout = 2000) => {
  const deadline = Date.now() + timeout;

  while (Date.now() < deadline) {
    const logs = await findLogsForModel(action);
    if (logs.length === 1) return logs[0];

    await new Promise((resolve) => setTimeout(resolve, 50));
  }

  throw new Error(`Expected ${action} audit log within ${timeout}ms`);
};
```

Use this for the positive create/delete assertions. The negative read assertion still needs either a complete observation window or, preferably, a deterministic way to await subscriber completion.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+  const callTool = async (
+    accessKey: string,
+    name: string,
+    args: Record<string, unknown>
+  ): Promise<JsonRpcResponse> => {
+    const res = await mcpRpc(accessKey, 'tools/call', { name, arguments: args });
+    return parseMcpResponse(res);
+  };
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts]: **P3 · Confidence: confirmed**

This block repeats response parsing, JSON-RPC transport, initialization, token creation, and tool-call helpers from the existing MCP API suites. Drift already exists: this copy checks initialization HTTP status but not JSON-RPC success or notification status.

Please extract a shared MCP API-test harness before adding more suites, or track that extraction as an explicit follow-up.

---
_AI assisted_

**Review by nclsndr — CHANGES_REQUESTED** [review]
Thank you for the PR @gonbaum 🙌 

Looks good on production code. Only findings on tests 👇 

--- 

Reviewed SHA: `62d9b2f7a294346c8fa0696d400bad6af8084d83`

Verdict: fix the P2 synchronization issue before merge; the P3 helper extraction may be tracked as an explicit follow-up.

Blast radius: four files across core, EE admin audit logs, and API tests. Runtime path spans MCP admin-token authentication, Koa request context, Document Service post-commit events, and the EE audit-log subscriber.

Published findings:
- P2 — Fixed sleep makes the regression test timing-dependent.
- P3 — MCP API-test infrastructure is duplicated again.

Unanchored findings:
- None.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 1 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts]: Fixed in `fd6921b907`. Replaced the fixed 200ms sleep with a bounded `waitForAuditLog` helper that polls for the expected row and fails after a finite timeout, used for the create/delete assertions.

For the negative read case, rather than a fixed observation window I made it deterministic: the read runs before a create whose audit log we await — once that create's log lands, the subscriber has processed everything before it in order, so the read having produced nothing is a reliable check.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 2 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts]: Extracted the shared MCP transport/JSON-RPC helpers (response parsing, session init, tool calls) into `tests/api/core/mcp/utils/mcp-client.ts` and migrated this suite to use it, so it no longer duplicates them. Also closed the drift you flagged: `initializeSession` now asserts JSON-RPC success and the notification status, not just the init HTTP status.

I kept the scope to this suite — migrating the existing MCP suites onto the shared client is tracked as a follow-up: https://linear.app/strapi/issue/EE-94/migrate-existing-mcp-api-test-suites-to-the-shared-mcp-client-harness

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+      data: { title: 'seed for read' },
+    });
+
+    const token = await createAdminToken([
+      fieldPermission(CM_ACTIONS.read, ['title']),
+      fieldPermission(CM_ACTIONS.create, ['title']),
+    ]);
+    await mcp.initializeSession(token.accessKey);
+
+    const read = await mcp.callTool(token.accessKey, `list_${SLUG}`, {});
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts]: **An empty or failed read can pass this test (P2)**

`callTool` discards HTTP status, and an empty response parses to `{}`. That value satisfies both assertions here, so an empty or failed read can pass this test.

The later create is not a deterministic completion barrier either: post-commit callbacks are launched without awaiting them, and independent EventHub emissions have no global ordering.

Please assert successful HTTP/JSON-RPC completion and returned list data, then observe `entry.*` emission or another explicit subscriber-completion signal instead of relying on the later create.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+      data: { title: 'created via mcp' },
+    });
+
+    expect(response.error).toBeUndefined();
+    expect(response.result?.isError).not.toBe(true);
+    const stored = await strapi.documents(MODEL_UID as UID.CollectionType).findMany({});
+    expect(stored).toHaveLength(1);
+    expect(stored[0].title).toBe('created via mcp');
+
+    const log = await waitForAuditLog('entry.create');
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts]: **The test does not verify the audit actor is the token owner (P2)**

The feature's audit invariant is that the stored actor is the authenticated Admin API-token owner, but this test checks only `payload.origin`. The handler unit test also does not assert either new `ctx.state` field, so a wrong truthy actor ID would remain green.

Please assert the audit row's `userId` equals the token owner's admin-user ID, and directly assert `ctx.state.user` plus `ctx.state.auditSource` in the handler unit test.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+  // The delete action carries no `fields` (passing one is rejected with a 400).
+  const actionPermission = (action: string): AdminPermission => ({
+    action,
+    subject: MODEL_UID,
+    conditions: [],
+    properties: {},
+  });
+
+  // entry.* events fire on a microtask after commit, so poll rather than sleep.
+  const waitForAuditLog = async (action: string, timeout = 2000) => {
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts:151]: **Polling no longer verifies that exactly one audit row is written (P3)**

This returns as soon as the first matching row appears. The create/delete tests then inspect that row but never verify there is exactly one audit row, so duplicate subscription or emission now passes; the previous assertions enforced a one-row result.

Please preserve bounded polling, then assert the settled matching-row collection has length one.

---
_AI assisted_

**Review by nclsndr — COMMENTED** [review]
Thank you for the follow-up @gonbaum!

Did not try to repro but looks legit from reading. Feel free to discard hallucinations (new review process on training)

---

Blast radius: MCP Admin API-token identity propagation, EE audit-log persistence, and MCP API-test synchronization. I left three inline findings: the read test can pass on an empty or failed response and relies on unsupported event ordering; actor attribution is not asserted; and polling no longer verifies exactly one audit row.

The synchronization thread should remain unresolved. The shared-helper extraction is now real, with older-suite migration tracked in EE-94, so that original thread can be resolved.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 3 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts]: Fixed in `4acf20c7b8`.

On the read passing silently: `callTool` now checks the call actually succeeded (200 + a real JSON-RPC envelope), and the test checks the list came back with rows. So an empty or failed read can't slip through anymore.

On the barrier — you were right, the later create didn't prove anything. Changed it to an actual signal: the test subscribes to the eventHub itself, and since subscribers run in order and each is awaited, my subscriber runs right after the audit-logs one. So by the time it sees the sentinel's `entry.create`, the row is already written. No more leaning on ordering between separate emits.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 4 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts]: Fixed in `4acf20c7b8`. The create/delete tests now assert the logged actor is the token owner (`log.user.id`), and the handlePost unit test asserts `ctx.state.user` and `ctx.state.auditSource` directly.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 5 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts:151]: Fixed in `4acf20c7b8`. `expectExactlyOneLog` still polls, but then waits a short settle window and asserts exactly one matching row — so a duplicate would fail it.

**Comment by gonbaum** [comment]
Thanks for the thorough review @nclsndr. the sync/ordering points were spot on and the tests are in much better shape for it. Pushed the fixes in `4acf20c7b8`, replied inline on each thread. One small heads-up on the exactly-one check left in that thread.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 5 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts:151]: One thing worth being upfront about: the settle window is time-bounded, so it only catches a duplicate that lands within ~200ms of the first row — it's a best-effort tripwire, not a hard proof of "exactly one". I went with that because there's no duplicate source today (single audit subscriber, and a `draftAndPublish: false` create emits `entry.create` once), so a real double-write would come from the same emit loop and land well inside the window. If you'd rather have a stricter guarantee here, I'm open to it — but it'd mean a longer fixed wait, which trades determinism speed for a mostly-theoretical case.

File: packages/core/core/src/services/mcp/handlers/__tests__/handlePost.test.ts [hunk 6]
```diff
@@ -247,6 +247,37 @@ describe('handlePost', () => {
...
+    StreamableHTTPServerTransport.mockImplementation(() => ({
+      handleRequest: jest.fn().mockResolvedValue(undefined),
+    }));
+
+    const handler = createPostHandler(deps);
+    const ctx = makeCtx(makeReq(), { headersSent: false } as unknown as ServerResponse);
+
+    await handler(ctx, () => Promise.resolve());
+
+    expect(ctx.state.user).toEqual({ id: 1 });
```
Comment (nclsndr) [inline packages/core/core/src/services/mcp/handlers/__tests__/handlePost.test.ts]: **The handler unit test cannot fail on a wrong actor (P2)**

`ctx.state.user` is asserted against `{ id: 1 }`, but the auth mock returns the same id for the token and for its owner (`credentials: { id: 1 }`, `user: { id: 1 }`), so this assertion cannot tell the token owner apart from the token itself — which is exactly the distinction the feature depends on.

Repro on this head: in `handlePost.ts`, change `ctx.state.user = authResult.user` to `ctx.state.user = authResult.credentials` — a real attribution bug — then run `yarn jest --config jest.config.js packages/core/core/src/services/mcp/handlers/__tests__/handlePost.test.ts`. Expected: this test fails. Actual: 13/13 still pass.

Could you give the mock distinct ids (e.g. `credentials: { id: 7 }`, `user: { id: 42 }`) and assert `{ id: 42 }`? That pins down which field is being read.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    expect(response.error).toBeUndefined();
+    expect(response.result?.isError).not.toBe(true);
+    const stored = await strapi.documents(MODEL_UID as UID.CollectionType).findMany({});
+    expect(stored).toHaveLength(1);
+    expect(stored[0].title).toBe('created via mcp');
+
+    const log = await expectExactlyOneLog('entry.create');
+    expect(log.payload).toMatchObject({ origin: 'mcp' });
+    // The audited actor is the admin token's owner.
+    expect(log.user.id).toBe(await getTokenOwnerId());
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts]: **The audited-actor assertion passes for any admin, not only the token owner (P2)**

`getTokenOwnerId()` resolves `admin@strapi.io`, which is the only admin the API-test harness seeds (`packages/utils/api-tests/strapi.js:99-101`, credentials at `:10-15`), and this suite creates no others. With a single admin row, the assertion holds for any implementation that merely lands on "some admin".

Repro: replace the actor lookup in `lifecycles.ts` with the first admin row (`strapi.db.query('admin::user').findOne({ orderBy: { id: 'asc' } })`), or hardcode `userId: 1`. Both are real attribution bugs. Expected: this test fails. Actual: it stays green, because that row *is* `admin@strapi.io`.

Could you create a second admin, issue the Admin API Token as that admin (`createAuthRequest({ strapi, userInfo: secondAdmin })` — the pattern already exists in `tests/api/core/admin/admin-admin-token-crud.test.api.ts:72-88`), and assert the audit row points at that second admin? Ownership does track the creator (`packages/core/admin/server/src/controllers/admin-token.ts:70` sets `adminUserOwner: ctx.state.user.id`), so the assertion becomes discriminating as soon as the owner isn't the only candidate. Same applies to the delete test at line 279.

---
_AI assisted_

File: packages/core/admin/ee/server/src/audit-logs/services/lifecycles.ts [hunk 8]
```diff
@@ -78,13 +78,17 @@ const createAuditLogsLifecycleService = (strapi: Core.Strapi) => {
...
     const isUsingAdminAuth = requestState?.route.info.type === 'admin';
+    const auditSource = requestState?.auditSource;
+    const isMcpAdminAction = auditSource === 'mcp';
     const user = requestState?.user;
-    if (!isUsingAdminAuth || !user) {
+    if ((!isUsingAdminAuth && !isMcpAdminAction) || !user) {
       return null;
     }
 
+    const origin = auditSource ?? 'admin';
```
Comment (nclsndr) [inline packages/core/admin/ee/server/src/audit-logs/services/lifecycles.ts]: **`origin` is derived from an undeclared, free-form state field (P3)**

`ctx.state.auditSource` isn't declared anywhere: it's absent from the request-state typing, and its only occurrences in the repo are this read, the write in `handlePost.ts:46`, and the test. Because `origin` falls back with `auditSource ?? 'admin'` while the recording gate accepts only `'mcp'` (line 84), a future caller setting any other value would get it stamped verbatim into every audit payload without the gate agreeing — and the compiler wouldn't flag it.

Would you consider a declared union (e.g. `type AuditSource = 'admin' | 'mcp'`) with the request-state field declared alongside it, so the gate and the label stay exhaustive together?

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+      const [log] = await findLogsForModel(action);
+      if (log) return log;
+      await sleep(50);
+    }
+    throw new Error(`Expected a ${action} audit log within ${timeout}ms`);
+  };
+
+  // Assert exactly one row landed, catching a late duplicate from a second
+  // subscription/emission that arrives after the first row appears.
+  const expectExactlyOneLog = async (action: string) => {
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts:163]: **Four overlapping ways to wait for the same audit write (P3)**

This file now carries `waitForAuditLog`'s polling (line 139), `expectExactlyOneLog`'s poll-plus-fixed-settle (here), the `quiesceAuditLogs` delete/sleep/recheck loop in `afterEach` (line 161), and the eventHub sentinel (line 230).

The sentinel is the strongest of the four, and it does hold: `emit()` awaits subscribers in registration order (`event-hub.ts:37-41`), the audit subscriber is registered at boot (`ee/server/src/index.ts:47-54` → `lifecycles.ts:158`), and `saveEvent` awaits its DB write with no deferral (`audit-logs.ts:36-45`). Using it for the positive create/delete cases too would remove `waitForAuditLog` and the fixed settle sleep, and would let `afterEach` become a plain delete, since each test would already have awaited its own audit write.

To be explicit: this wouldn't strengthen the duplicate check. Ruling out a later duplicate needs an unbounded wait, so your reasoning in the exactly-one thread stands and the tripwire is a fair trade. This is only about collapsing four mechanisms into one.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+  const findLogsForModel = async (action?: string) => {
+    const logs = await strapi.db.query('admin::audit-log').findMany({
+      ...(action ? { where: { action } } : {}),
+      populate: ['user'],
+    });
+    return logs.filter((log: { payload?: { uid?: string } }) => log.payload?.uid === MODEL_UID);
+  };
+
+  // Admin tokens are owned by the super admin creating them via createAuthRequest.
+  const getTokenOwnerId = async () => {
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts]: **Prefer the harness's superadmin helpers over a hardcoded email (nit)**

The harness exports `superAdmin.credentials` (`packages/utils/api-tests/strapi.js:108-111`) and the agent exposes `rq.getLoggedUser()` (`packages/utils/api-tests/agent.js:143-145`), used that way in `tests/api/core/admin/admin-user.test.api.js:58-63`. Either is sturdier than the inline `'admin@strapi.io'`.

One small correction to the comment as well: token ownership follows whichever authenticated admin creates the token (`packages/core/admin/server/src/controllers/admin-token.ts:70`), not the super admin specifically. It happens to be the super admin here only because that's who `createAuthRequest` logs in as.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    await mcp.initializeSession(token.accessKey);
+
+    // A test subscriber registered now runs after the audit-logs subscriber in
+    // the emit loop, so when it sees the sentinel's entry.create the audit row
+    // is already committed — an explicit completion signal, not a timed guess.
+    let onSentinelAudited!: () => void;
+    const sentinelAudited = new Promise<void>((resolve) => {
+      onSentinelAudited = resolve;
+    });
+    const unsubscribe = strapi.eventHub.subscribe(async (name: string, ...args: any[]) => {
```
Comment (nclsndr) [inline tests/api/core/mcp/mcp-audit-logs.test.api.ts:244]: **Worth recording the assumption this barrier rests on (nit)**

The ordering argument holds today, but it inverts if the audit subscriber ever re-registers after this one: `register()` re-pushes `handleEvent` to the end of the subscriber array (`lifecycles.ts:127-141` → `:158`), which both `ee.enable` and `ee.update` trigger. The sentinel would then resolve before the row is written.

Nothing emits those events mid-test and EE is enabled at boot, so it isn't reachable in this suite — a one-line comment noting the assumption would just save the next reader from re-deriving it.

---
_AI assisted_

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 7 above)
Comment (nclsndr) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts]: Setup and runtime might be fragile. If happens to feel too hacky we can wait test setup refactor work to land. Leaving a comment on the flaw only.

**Review by nclsndr — DISMISSED** [review]
Thanks for turning these around so quickly, @gonbaum.

Blast radius: `@strapi/core`'s MCP request context (`POST /mcp` now sets `ctx.state.user` and `ctx.state.auditSource`), EE audit-log persistence (every audit row now carries `payload.origin`), and the shared MCP API-test client.

On the two synchronization points from last round: both hold up. The `callTool` transport assertion closes the empty/failed-read gap, and the eventHub sentinel is a real completion signal — `emit()` awaits subscribers in registration order, the audit subscriber is registered at boot, and `saveEvent` awaits its write directly, so the row is committed by the time your subscriber fires. I also checked that the suite genuinely executes in EE CI rather than being skipped for want of a license (`PASS tests/api/core/mcp/mcp-audit-logs.test.api.ts` in the `[EE] API Integration (mysql, node 22, shard 2/5)` job). Your reading of the exactly-one tripwire is right too — a hard guarantee there needs an unbounded wait.

The one thing I'd still like closed is actor attribution. Assertions now exist on both sides, but neither can go red: the unit mock uses the same id for the token and for its owner, and the API test compares against the only admin the harness seeds. Both have concrete repros inline.

The rest are non-blocking: `auditSource` is an undeclared free-form field whose looseness `origin` inherits, the test file now has four overlapping ways to wait for the same write, and two small nits.

Unanchored findings:

- The contributor docs for audit logs are now stale in two places: `docs/docs/docs/01-core/admin/01-ee/02-audit-logs.md:62` still says the guard only records events coming from admin requests, and the documented `Event` shape at `:47-54` omits the newly always-present `payload.origin`.
- Related: `docs/docs/docs/01-core/strapi/event-hub.md:34` documents only that `emit` resolves once all subscribers have run — not the registration-order sequencing this PR's test now depends on. Since shipped test code relies on it, that guarantee is worth writing down.

---
_AI assisted_

**Comment by nclsndr** [comment]
adding @innerdvations to follow up for next week if necessary

File: packages/core/core/src/services/mcp/handlers/__tests__/handlePost.test.ts (hunk 6 above)
Comment (gonbaum) [reply packages/core/core/src/services/mcp/handlers/__tests__/handlePost.test.ts]: Fixed in `90f1d21aa5`. The auth mock now returns distinct ids (`credentials: { id: 7 }`, `user: { id: 42 }`) and the test asserts `ctx.state.user` is `{ id: 42 }` — so swapping `authResult.user` for `authResult.credentials` now fails it.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 7 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts]: Fixed in `90f1d21aa5`. The suite now creates a dedicated second admin, issues the token as that admin, and asserts the audit row's `user.id` is that admin's id — so it no longer holds for "some admin". Same for the delete test.

File: packages/core/admin/ee/server/src/audit-logs/services/lifecycles.ts (hunk 8 above)
Comment (gonbaum) [reply packages/core/admin/ee/server/src/audit-logs/services/lifecycles.ts]: Fixed in `90f1d21aa5`. Added a shared `type AuditSource = 'admin' | 'mcp'` in `@strapi/types` and typed the `origin` label against it, so the gate and the label stay exhaustive together. One caveat worth naming: I also declared `auditSource` on Koa's `DefaultState`, but since that interface is `any`-based the augmentation is really just documentation — it won't catch a bad value at the write site. The real enforcement is at the read site in `lifecycles.ts` where `origin` is typed `AuditSource`.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 9 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts:163]: Left as-is for now. The sentinel is the strongest signal, agreed, but as you noted collapsing them wouldn't strengthen the duplicate check — and `expectExactlyOneLog`'s settle window is doing real work there. Happy to consolidate onto the sentinel in a follow-up if you'd prefer; didn't want to expand this PR's test churn further.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 10 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts]: Resolved as a side effect of the actor fix: there's no hardcoded `admin@strapi.io` anymore — the token owner is the second admin created via `createUtils`, and the assertion reads its id directly. Also thanks for the correction on ownership following the token creator, not the super admin specifically — that's reflected now.

File: tests/api/core/mcp/mcp-audit-logs.test.api.ts (hunk 11 above)
Comment (gonbaum) [reply tests/api/core/mcp/mcp-audit-logs.test.api.ts:244]: Done in `90f1d21aa5` — added a comment noting the ordering assumption (that the audit subscriber isn't re-registered mid-test via ee.enable/update) and why it holds here.

**Comment by gonbaum** [comment]
Thanks Nico — good catches on the actor assertions especially. All addressed in `90f1d21aa5`: distinct ids in the unit mock, a dedicated second admin so the actor assertion is discriminating, and a shared `AuditSource` union. Left the four-wait-mechanisms nit as-is with a note (happy to consolidate in a follow-up), and flagged inline that the `DefaultState` augmentation is documentation-only given the `any` base. Rebased on develop to clear the conflict.

File: packages/core/admin/ee/server/src/audit-logs/services/lifecycles.ts (hunk 8 above)
Comment (gonbaum) [reply packages/core/admin/ee/server/src/audit-logs/services/lifecycles.ts]: Update in `3e1174e6d3` — I ended up removing the `DefaultState` augmentation entirely: it conflicted with webpack-dev-middleware's state type in `@strapi/strapi` and broke the build, and since Koa's `DefaultState` is `any`-based it wasn't buying real type safety anyway. The shared `AuditSource` union stays and is still enforced where it's read in `lifecycles.ts` (`origin: AuditSource`). Also fixed a test-isolation bug the change surfaced (`93b9f2c756`): the second admin wasn't cleaned up, so reruns on a reused DB made the actor assertion flaky — now deleted in `afterAll`.

**Review by innerdvations — APPROVED** [review]
Looks good now! Did the following manual tests:

```
Happy path
1. MCP initialize + notifications/initialized — PASS
2. tools/list includes create_temp / list_temp / update_temp / delete_temp — PASS
3. Admin CM create (name=via-admin-panel) → audit entry.create with payload.origin=admin — PASS
4. MCP create_temp (name=via-mcp-create) → entry.create origin=mcp — PASS
5. MCP create actor is token owner (user.id=3) — PASS
6. MCP list_temp returns rows — PASS
7. MCP list_temp adds 0 new entry.* audit rows; sentinel create adds exactly +1 origin=mcp — PASS
8. MCP update_temp → entry.update origin=mcp — PASS
9. MCP delete_temp → entry.delete origin=mcp — PASS
10. MCP delete actor is token owner (user.id=3 / test@strapi.io) — PASS
Edge cases
11. Bad bearer on /mcp → 401 Authentication required — PASS
12. Content-API (full-access) token on /mcp → 401 Authentication required — PASS
13. Read-only admin token (explorer.read only) create_temp → isError, tool disabled — PASS
```
